dali/operators/image/paste/multipaste.h

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_IMAGE_PASTE_MULTIPASTE_H_
+#define DALI_OPERATORS_IMAGE_PASTE_MULTIPASTE_H_
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/operator/arg_helper.h"
+#include "dali/core/format.h"
+#include "dali/util/crop_window.h"
+#include "dali/pipeline/data/types.h"
+
+
+namespace dali {
+
+template <typename Backend>
+class MultiPasteOp : public Operator<Backend> {
+ public:
+  ~MultiPasteOp() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(MultiPasteOp);
+
+ protected:
+  using Coords = TensorView<typename compute_to_storage<Backend>::type, const int, 1>;
+
+  explicit MultiPasteOp(const OpSpec &spec)
+      : Operator<Backend>(spec)
+      , output_type_arg_(spec.GetArgument<DALIDataType>("dtype"))
+      , output_type_(DALI_NO_TYPE)
+      , input_type_(DALI_NO_TYPE)
+      , output_size_("output_size", spec)
+      , in_idx_("in_ids", spec)
+      , in_anchors_("in_anchors", spec)
+      , in_shapes_("shapes", spec)
+      , out_anchors_("out_anchors", spec) {
+    zero_anchors_ = Coords(zeros_, {2});
+
+    if (std::is_same<Backend, GPUBackend>::value) {
+      kernel_manager_.Resize(1, 1);
+    } else {
+      kernel_manager_.Resize(num_threads_, max_batch_size_);
+    }
+  }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  template<int ndim = 2>
+  bool Intersects(const Coords& anchors1, const Coords& shapes1,
+                  const Coords& anchors2, const Coords& shapes2) const {
+    for (int i = 0; i < ndim; i++) {
+      if (anchors1.data[i] + shapes1.data[i] <= anchors2.data[i]
+          || anchors2.data[i] + shapes2.data[i] <= anchors1.data[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void AcquireArguments(const OpSpec &spec, const workspace_t<Backend> &ws) {
+    const auto &images = ws.template InputRef<Backend>(0);
+
+    auto curr_batch_size = ws.GetRequestedBatchSize(0);
+    if (curr_batch_size == 0)
+      return;
+
+    output_size_.Acquire(spec, ws, curr_batch_size, true);
+    in_idx_.Acquire(spec, ws, curr_batch_size, false);
+    if (out_anchors_.IsDefined()) {
```

Comment:
This is a suggestion so that you don't need to do manual validation below:
```
TensorListShape<2> expected_sh(curr_batch_size);
for (int i = 0; I < curr_batch_size; i++) {
  expected_sh.set_tensor_shape(i, TensorShape<2>{in_idx_[i].shape[0], static_ndim});
}
if (out_anchors_.IsDefined()) {
  out_anchors_.Acquire(spec, ws, curr_batch_size, expected_sh);
}
if (in_anchors_.IsDefined()) {
  in_anchors_.Acquire(spec, ws, curr_batch_size, expected_sh);
}
if (in_shapes_.IsDefined()) {
  in_shapes_.Acquire(spec, ws, curr_batch_size, expected_sh);
}
```
If you do this, the shape is already enforced to be ``expected_sh`` so you can remove the validation done below (lines 120-136).

Reply:
It seems this only works when enforcing uniform shapes, the last argument of this variant of Acquire is a const TensorShape<ndim>, not a TensorListShape

Reply:
You are right, my mistake